### PR TITLE
Nixpkgs config: NIXPKGS_CONFIG -> <nixpkgs-config>

### PR DESCRIPTION
**Pull request by michaelpj**
###### Motivation for this change
Fixes https://github.com/NixOS/nixpkgs/issues/28085.

- Expand documentation for nixpkgs config in a similar fashion to https://github.com/NixOS/nixpkgs/pull/28082
- Throw an error if potential config files are not files as expected.
- Change `NIXPKGS_CONFIG` to `<nixpkgs-config>` throughout.

I believe this is behaviour preserving except for:
- Anyone who consumes `NIXPKGS_CONFIG` (not clear why you would do this)
- Anyone who separately sets `NIXPKGS_CONFIG`.
- Anyone who clears their `NIX_PATH` on a NixOS system rather than extending it. In particular, users of the `nix.nixPath` option.

Possibly we could make things easier for the second set of people by continuing to honour `NIXPKGS_CONFIG` for a while with a warning, but I'm not sure if it's worth it.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by michaelpj** [comment]
Note that contra my claim in https://github.com/NixOS/nixpkgs/issues/28085, adding this to `nix.nixPath` *does* put it in the environment for users too.

**Comment by vyp** [comment]
What does this do for users of `nix.nixPath`?

**Comment by michaelpj** [comment]
Setting `nix.nixPath` overwrites the default, so if you want any of the default bits you have to copy them over yourself. This adds to the default value, so anyone using it won't get that, where they probably would have got `NIXPKGS_CONFIG`.

I'm not sure there's much we can do about this, short of making `nix.nixPath` configurable as an attribute set which could then be merged with the default. However, I don't think it's a bit problem, since it seems like almost nobody uses `NIXPGS_CONFIG`, since it's undocumented.

**Comment by vyp** [comment]
That's what I thought, thank you for explaining.

File: pkgs/top-level/impure.nix [hunk 1]
```diff
@@ -24,16 +24,26 @@ in
...
+      if configFile != "" && pathExists configFile then
+        if isDir configFile then 
+          throw (configFile + " must be a file")
+        else import configFile
+      else if configFile2 != "" && pathExists configFile2 then
+        if isDir configFile2 then 
+          throw (configFile2 + " must be a file")
+        else import configFile2
+      else if configFile3 != "" && pathExists configFile3 then 
+        if isDir configFile then
```
Comment (fpletz) [inline pkgs/top-level/impure.nix]: Shouldn't this be `configFile3`?

Comment (michaelpj) [inline pkgs/top-level/impure.nix]: Ouch, good spot.

Comment (michaelpj) [inline pkgs/top-level/impure.nix]: Fixed.

**Comment by michaelpj** [comment]
I've also added a nixos release note, since this is technically a breaking change.

**Comment by fpletz** [comment]
I like the change but I'm not sure if this should make 17.09. Has anybody tested this yet?

**Comment by michaelpj** [comment]
Yeah, I was somewhat putting this up for discussion - it only really makes things more consistent, rather than bringing any concrete improvement.

I think putting this in after the branch-off would make sense. I'll update it and test it a bit more when that happens.

**Comment by michaelpj** [comment]
Do we think this is still desirable? I think so, but I don't feel strongly about it

**Comment by matthewbauer** [comment]
Pinging the status on this. Is this something we really want? We may be able to do this for 18.09

**Comment by Infinisil** [comment]
Closing this because it doesn't seem to be something anybody really needs right now. This PR looks good otherwise, and I'll be happy to merge this in the future when people want it, it's well tested, and has a warning for backwards compatibility (and a bit of a refactor for the config selection).

**Pull request closed without merging**